esp-hal/src/i2s/master/camera.rs

```diff
@@ -0,0 +1,404 @@
+        }
+
+        // set I2S_RX_STOP_EN if needed
+
+        // start: set I2S_RX_START
+        self._i2s.rx_start(len);
+        Ok(())
+    }
+
+    /// Starts a DMA transfer to receive data from the camera peripheral.
+    pub fn receive<BUF: DmaRxBuffer>(
+        mut self,
+        mut buf: BUF,
+        len: usize,
+    ) -> Result<CameraTransfer<'d, I2S, BUF>, Error> {
+        self.start_rx_transfer(&mut buf, len)?;
+        Ok(CameraTransfer {
+            camera: ManuallyDrop::new(self),
+            buffer_view: ManuallyDrop::new(buf.into_view()),
+        })
+    }
+}
+
+/// Represents an ongoing (or potentially stopped) transfer from the Camera to a
+/// DMA buffer.
+pub struct CameraTransfer<'d, I2S: DmaEligible + RegisterAccess, BUF: DmaRxBuffer> {
+    camera: ManuallyDrop<Camera<'d, I2S>>,
+    buffer_view: ManuallyDrop<BUF::View>,
+}
+
+impl<'d, I2S: RegisterAccess, BUF: DmaRxBuffer> CameraTransfer<'d, I2S, BUF> {
+    /// Returns true when [Self::wait] will not block.
+    pub fn is_done(&self) -> bool {
+        self.camera.rx_channel.has_dscr_empty_error() // IN_DSCR_EMPTY (i.e. No more buffer space)
+            || self.camera.rx_channel.has_error() // IN_DSCR_ERR (i.e. bad
+                                                  // descriptor)
+    }
+
+    /// Stops this transfer on the spot and returns the peripheral and buffer.
+    pub fn stop(mut self) -> (Camera<'d, I2S>, BUF) {
+        self.stop_peripherals();
+        let (camera, view) = self.release();
+        (camera, BUF::from_view(view))
+    }
+
+    /// Waits for the transfer to stop and returns the peripheral and buffer.
+    ///
+    /// Note: The camera doesn't really "finish" its transfer, so what you're
+    /// really waiting for here is a DMA Error. You typically just want to
+    /// call [Self::stop] once you have the data you need.
+    pub fn wait(mut self) -> (Result<(), DmaError>, Camera<'d, I2S>, BUF) {
+        while !self.is_done() {}
+
+        // Stop the DMA as it doesn't know that the camera has stopped.
+        self.camera.rx_channel.stop_transfer();
+
+        // Note: There is no "done" interrupt to clear.
+
+        let (camera, view) = self.release();
+
+        let result = if camera.rx_channel.has_error() {
+            Err(DmaError::DescriptorError)
+        } else {
+            Ok(())
+        };
+
+        (result, camera, BUF::from_view(view))
+    }
+
+    fn release(mut self) -> (Camera<'d, I2S>, BUF::View) {
+        // SAFETY: Since forget is called on self, we know that self.camera and
+        // self.buffer_view won't be touched again.
+        let result = unsafe {
+            let camera = ManuallyDrop::take(&mut self.camera);
+            let view = ManuallyDrop::take(&mut self.buffer_view);
+            (camera, view)
+        };
+        core::mem::forget(self);
+        result
+    }
+
+    fn stop_peripherals(&mut self) {
+        self.camera._i2s.rx_stop();
```

Comment:
the DMA needs to be stopped here too

Reply:
does the order matter? should i stop DMA first or I2S RX?

Reply:
Do what the other rx drivers do